### PR TITLE
Add basic Rotary Embeddings layer (Su et al., 2021)

### PR DESCRIPTION
**Pull request by danieldk**
## Description

Add the rotary embeddings layer that is used by various GPT-based LLMs.

### Types of change

Enhancement.

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: curated_transformers/models/pytorch/embeddings.py [hunk 1]
```diff
@@ -28,3 +29,57 @@ def forward(self, x: Tensor) -> Tensor:
...
+            Rotary embedding dimensionality.
+        seq_len (int):
+            Number of positons to initially pre-cache.
+        base (int):
+            The base used for Θ_i, determines the cycle length of the
+            embeddings."""
+        super().__init__()
+
+        # Θ_i = 10000^(-2(i-1)/d)
+        theta = torch.pow(base, -torch.arange(0, width, 2, dtype=torch.float) / width)
```
Comment (shadeMe) [inline curated_transformers/models/pytorch/embeddings.py:55]: We should add a check to ensure that `width` is even (or else we break at the `view()` call in line 68).

File: curated_transformers/models/pytorch/embeddings.py [hunk 2]
```diff
@@ -28,3 +29,57 @@ def forward(self, x: Tensor) -> Tensor:
...
+
+        self._create_rotary_embed(width=width, length=seq_len)
+
+    def _create_rotary_embed(self, *, width: int, length: int):
+        # mΘ
+        position = torch.arange(length).unsqueeze(1)
+        m_theta = position * self.theta.unsqueeze(0)
+
+        # We apply both sin and cos twice (see Eq 15, 34).
+        m_theta = torch.cat([m_theta, m_theta], dim=-1)
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/models/pytorch/embeddings.py:66]: Yeah, that's what the paper says, but then the real-world implementation doesn't, so we have to do the same thing to be compatible. I can add a comment.

**Comment by danieldk** [comment]
Putting this in draft temporarily. There are some bit where I originally thought that they need to be in the attention layer, that can also move to this layer.

**Comment by danieldk** [comment]
Closing this one and making a new PR, since stuff moved around and there are some changes, it's easier to make a new PR. I'll integrate the comments as well.

**Pull request closed without merging**